### DISCUSSION TITLE
Should I avoid the union expression syntax with `|` when declaring response_model's?

### DISCUSSION THREAD

**Discussion by julianolf** [fastapi/fastapi#11434] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class BaseItem(BaseModel):
    description: str
    type: str


class CarItem(BaseItem):
    type: str = "car"


class PlaneItem(BaseItem):
    type: str = "plane"
    size: int


items = {
    "item1": {"description": "All my friends drive a low rider", "type": "car"},
    "item2": {
        "description": "Music is my aeroplane, it's my aeroplane",
        "type": "plane",
        "size": 5,
    },
}


@app.get("/items/{item_id}", response_model=PlaneItem | CarItem)
async def read_item(item_id: str):
    return items[item_id]
```

### Description

According to the user's guide tutorial this code should **NOT** work, but it does. I tried different versions of Python (from 3.10.0 to 3.12.1), everything works just fine, even the auto generated docs looks correct.

> But if we put that in `response_model=PlaneItem | CarItem` we would get an error, because Python would try to perform an invalid operation between `PlaneItem` and `CarItem` instead of interpreting that as a type annotation.

https://fastapi.tiangolo.com/tutorial/extra-models/#union-in-python-310

Considering that I'm working on the latest version of Python, is there any caveats that I should be aware of and avoid using the `|` syntax when defining unions in responde_model's?

### Operating System

macOS

### Operating System Details

macOS Sonoma 14.4.1

### FastAPI Version

0.110.1

### Pydantic Version

2.7.0

### Python Version

Python 3.10.0

### Additional Context

Looking at the REPL I can see there's a slightly diffence between the two returned types, but it is my understading of Python's documentation that they should behave the same.
```python
>>> from pydantic import BaseModel
>>> class A(BaseModel):
...     pass
...
>>> class B(BaseModel):
...     pass
...
>>> arg1 = A | B
>>> arg1
__main__.A | __main__.B
>>> type(arg1)
<class 'types.UnionType'>
>>> isinstance(A(), arg1)
True
>>> isinstance(B(), arg1)
True
>>> from typing import Union
>>> arg2 = Union[A, B]
>>> arg2
typing.Union[__main__.A, __main__.B]
>>> type(arg2)
<class 'typing._UnionGenericAlias'>
>>> isinstance(A(), arg2)
True
>>> isinstance(B(), arg2)
True
>>>
```
I guessing there was an issue with this syntax back in the day when Python 3.10 came out but it was fixed and the documentation got outdated.

I just want to make sure of not crashing my applications due to syntax preferences.

**Comment by denis-samatov** [comment, marked as answer]
You can safely use `PlaneItem | CarItem` as the value of `response_model` on Python 3.10+.

PEP 604 made `A | B` produce a real runtime union object (`types.UnionType`). It is not limited to annotation syntax, so FastAPI and Pydantic can consume it exactly as they consume `typing.Union[A, B]`. The sentence saying that Python would perform an invalid operation is outdated; the code example on the same current documentation page already uses:

```python
@app.get(
    "/items/{item_id}",
    response_model=PlaneItem | CarItem,
)
async def read_item(item_id: str):
    return items[item_id]
```

The practical compatibility boundary is Python 3.10. Use `typing.Union` only if the code must still run on Python 3.9 or earlier.

There is a separate modeling caveat in the example: both `type` fields are plain `str` with defaults. For reliable variant selection, especially as the models grow, make them literal tags and use a discriminated union:

```python
from typing import Annotated, Literal

from fastapi import FastAPI
from pydantic import BaseModel, Field


class BaseItem(BaseModel):
    description: str


class CarItem(BaseItem):
    type: Literal["car"] = "car"


class PlaneItem(BaseItem):
    type: Literal["plane"] = "plane"
    size: int


Item = Annotated[
    PlaneItem | CarItem,
    Field(discriminator="type"),
]


app = FastAPI()


@app.get("/items/{item_id}", response_model=Item)
async def read_item(item_id: str):
    return items[item_id]
```

That makes the OpenAPI discriminator explicit and avoids relying on union-member ordering. Without a discriminator, keep the more specific model first, as the FastAPI documentation recommends.

Reference: https://fastapi.tiangolo.com/tutorial/extra-models/#union-or-anyof

*If this solves your issue, please consider marking this comment as the answer!*

**Reply by julianolf** [reply]
Thanks @denis-samatov.
